test/e2e/util/mongotester/mongotester.go

```diff
@@ -0,0 +1,237 @@
+// to connect to the MongoDB resource
+func (m *Tester) ConnectivitySucceeds(opts ...*options.ClientOptions) func(t *testing.T) {
+	return m.connectivityCheck(true, opts...)
+}
+
+// ConnectivitySucceeds performs a basic check that ensures that it is not possible
+// to connect to the MongoDB resource
+func (m *Tester) ConnectivityFails(opts ...*options.ClientOptions) func(t *testing.T) {
+	return m.connectivityCheck(false, opts...)
+}
+
+func (m *Tester) HasKeyfileAuth(tries int) func(t *testing.T) {
+	return m.hasAdminParameter("clusterAuthMode", "keyFile", tries)
+}
+
+func (m *Tester) HasFCV(fcv string, tries int) func(t *testing.T) {
+	return m.hasAdminParameter("featureCompatibilityVersion", map[string]interface{}{"version": fcv}, tries)
+}
+
+func (m *Tester) hasAdminParameter(key string, expectedValue interface{}, tries int) func(t *testing.T) {
+	return func(t *testing.T) {
+		if err := m.ensureClient(); err != nil {
+			t.Fatal(err)
+		}
+
+		database := m.mongoClient.Database("admin")
+		assert.NotNil(t, database)
+
+		found := false
+		for !found && tries > 0 {
+			<-time.After(10 * time.Second)
+			actualValue, err := m.getAdminSetting(key)
+			t.Logf("actualValue: %+v", actualValue)
+			if err != nil {
+				t.Logf("Unable to get admin setting %s with error : %s", key, err)
+				continue
+			}
+			found = reflect.DeepEqual(expectedValue, actualValue)
+			tries--
+		}
+		assert.True(t, found)
+	}
+}
+
+// getAdminSetting will get a setting from the admin database.
+func (m *Tester) getAdminSetting(key string) (interface{}, error) {
+	var result map[string]interface{}
+	err := m.mongoClient.
+		Database("admin").
+		RunCommand(context.TODO(), bson.D{{Key: "getParameter", Value: 1}, {Key: key, Value: 1}}).
+		Decode(&result)
+	if err != nil {
+		return nil, err
+	}
+	value := result[key]
+	return value, nil
+}
+
+func (m *Tester) connectivityCheck(shouldSucceed bool, opts ...*options.ClientOptions) func(t *testing.T) {
+	connectivityOpts := defaults()
+	return func(t *testing.T) {
+		ctx, cancel := context.WithTimeout(context.Background(), connectivityOpts.ContextTimeout)
+		defer cancel()
+
+		if err := m.ensureClient(opts...); err != nil {
+			t.Fatal(err)
+		}
+
+		attempts := 0
+		// There can be a short time before the user can auth as the user
+		err := wait.Poll(connectivityOpts.IntervalTime, connectivityOpts.TimeoutTime, func() (done bool, err error) {
+			attempts++
+			collection := m.mongoClient.Database(connectivityOpts.Database).Collection(connectivityOpts.Collection)
+			_, err = collection.InsertOne(ctx, bson.M{"name": "pi", "value": 3.14159})
+			if err != nil && shouldSucceed {
+				return false, nil
+			}
+			if err == nil && !shouldSucceed {
+				t.Logf("Was successfully able to connect, when we should not have been able to!")
+				return false, nil
+			}
+
+			successfulCheck := (err != nil && !shouldSucceed) || (err == nil && shouldSucceed)
```

Comment:
[q] Can this boolean ever be false?

```golang
err == nil && shouldSucceed -> successfulCheck == true
err == nil && !shouldSucceed -> we return at line 139
err != nil && shouldSucceed -> we return at line 136
err != nil && !shouldSucceed -> successfulCheck == true
```

Reply:
yes you're right, I think we can just remove this last check entirely